accepted/2023/preview-apis/preview-apis.md

```diff
@@ -0,0 +1,291 @@
+  these features is that they are in support of each other. For example, generic
+  math requires new runtime behaviors for interfaces, which the language has to
+  support, and of course new interfaces and methods in core library that
+  leverages them. If you want to use generic math, you need a combination of
+  runtime, framework, and language that were designed to work in conjunction.
+
+* **library-level preview feature**. These are library APIs that don't require
+  any platform-level preview features to be consumed. The only difference to
+  regular APIs is that these are in preview so their shape isn't stable and thus
+  offers no backwards compatibility when upgrading to a later version of the
+  library.
+
+Due to their nature, platform-level preview features can only be consumed in a
+specific combination of framework and SDK: namely, the framework and SDK version
+must match. For example, you can only use preview features in `net6.0` when
+using the .NET 6 SDK and only use preview features in `net7.0` when using the
+.NET 7 SDK. Specifically, you can't use the .NET 7 SDK to use preview features
+in `net6.0`. That's because preview features aren't backwards compatible and
+platform-level preview features span both framework and tooling. In
+multi-targeting scenarios this means you can only use preview features for the
+latest framework version.
+
+For platform-level preview features this restriction makes sense because it
+models reality. However, for library-level preview features not only does this
+restriction not make sense, it's highly undesirable. Library authors simply want
+a way to express "these parts of my API surface might change". This statement
+isn't tied to any runtime-, framework-, or language features.
+
+Another concern we had for platform-level preview features is that we wanted to
+ensure that application developers are aware when their applications have a
+dependency on them. That is, we want to avoid a case where a library on NuGet
+uses platform-level preview features as an implementation detail that the
+application consumer isn't aware of. So we made the opt-in mechanism viral: a
+library that turns on platform-level preview features requires its consumer to
+turn them on as well.
+
+Since library-level preview features cannot change how the .NET runtime itself
+operates, we don't have that same level of concern. In fact, we believe the
+viral nature is also undesirable.
+
+For the remainder of this proposal we'll refer to *library-level preview*
+features as *experimental APIs*.
+
+## Scenarios and User Experience
+
+### Shipping an experimental API in a stable package
+
+Martin builds a library that integrates with ASP.NET Core and provides features
+to make it easier to build services. The library is distributed as a NuGet
+package. The package is already in a version well past 1.0 and is considered
+stable.
+
+He ships the package often and doesn't want to maintain multiple flavors of the
```

Comment:
> doesn't want to maintain multiple flavors of the package, one marked as prerelease and one marked as stable.

I don't quite understand this. It feels like this proposal is a way to escape semantic versioning. If Martin does ship an experimental package it should be clear to consumers that things will change. If someone takes a dependency on a pre-release package and tries to ship a release package that depends on it, nuget will warn the user and they will need to explicitly opt out of this behavior. I don't see how shipping pre-release apis in release packages is the better solution from a user-confusion standpoint.

Reply:
Is the problem that we are solving for these nuget package authors that they want a better way to advertise their new functionality and asking people to consume pre-release packages is too big a burden? Why should we ever ship pre-release packages with this proposal?

Reply:
As we discussed this with the Azure SDK team, we came to realize that there are two dimensions:

1. General updates/refactorings/etc to a package, that may thus be of uncertain quality across the entire surface area.
2. Specific updates to one set of APIs that are in preview forms, without the stability risks implied to the rest of the package.

Today pre-release packages in NuGet do not have a way to distinguish between these two.  This is a hope to allow us give people a chance to use and give feedback on preview apis/services without taking on all of the risk inherent in 1.  We see that many of our customers can't use the new service features we build for them when they are in preview because they have company wide central package management, and policies forbidding the use of unstable packages.

Reply:
It feels like it would be cheaper to build this concept into nuget (`1.0.0-experimental` and `1.0.0-beta`) instead of taking compiler and other changes? I don't quite understand why we need and entirely new concept that is orthogonal to our existing ones here.

Reply:
Or without any platform changes, you could ship a new package `MyPackage.Experimental` like what ASP.NET did. I assume there was a lot of customer pain here from that approach and that's why it's not a viable solution?

Reply:
@jmarolf I think what has emerged here is evidence that there's a need (demonstrated eg by they're doing it already) for package authors to be able to include non-stable features within stable packages. Ie., they don't strictly sequence preview and stable work, because that isn't agile enough for them, and they and their customers don't want to maintain parallel package sequences either (`*Experimental*`..) as it's harder to reason about and doubles their costs. Second it is understood that in many cases those non stable features map closely to a new API (ie., if you use other API we can say you're using stable API in a strict sense). So this is not about getting around corporate policies that forbid preview packages but rather to make it possible for customers to work with the above behaviors in a safer way.

Reply:
On the producer side, of course it is a new build flavor and package series. On the consumer side -- @geeknoid can give a better answer perhaps. I'm imagining transitive dependencies may be part of the challenge.

Reply:
One of the big challenges with using separate packages is introducing experimental APIs on existing types. There’s no good solution that doesn’t lead to binary compatibility breaks in the future for APIs that don’t change shape (aka non-source-breaking changes), which is what the “separate package” model is generally used to support.

Reply:
> On the producer side, of course it is a new build flavor and package series

Unsure if the additional MSBuild scripts that would need to be added are more expensive than this feature, but I see your point.

> I'm imagining transitive dependencies may be part of the challenge.
> One of the big challenges with using separate packages is introducing experimental APIs on existing types. 

Thanks for clarifying, these are good examples of things that can't really be worked out via existing mechanisms.